### PR TITLE
use arduinoJson for fields

### PR DESCRIPTION
**Pull request by henrygab**
This should reduce heap fragmentation (by constant appending of small strings).

Uses arduinoJson to simplify (and make safer) the generation of JSON.  Uses custom converter for `field` structure to make easy to use.

Also explicitly makes most of the code in `fields.cpp` internal, to ensure safe to modify the internals.  Only public APIs are put outside the namespace.

There's also some minor changes (removing `auto` references in map.cpp) in anticipation of enabling ESP32.

### CODE REVIEW

File: esp8266-fastled-webserver/Fields.cpp [hunk 1]
```diff
@@ -17,479 +17,526 @@
...
-    if(field.getValue) {
-      if (field.type == ColorFieldType || field.type == StringFieldType || field.type == LabelFieldType) {
-        json += ",\"value\":\"" + field.getValue() + "\"";
-      }
-      else {
-        json += ",\"value\":" + field.getValue();
+    dst[F("name")]  = field.name;
+    dst[F("label")] = field.label;
+
+    dst[F("type")]  = ToString(field.type);
```
Comment (tobi01001) [inline esp8266-fastled-webserver/Fields.cpp:343]: If the modification to use enumeration in app.js as well is done, this needs to be change to:
`dst[F("type")]  = (uint8_t)field.type;`
```suggestion
    dst[F("type")]  = (uint8_t)field.type;
```

File: esp8266-fastled-webserver/Fields.cpp [hunk 2]
```diff
@@ -17,479 +17,526 @@
...
+  struct Field {
+    const String name;
+    const String label;
+    Field_t type;
+    uint8_t min;
+    uint8_t max;
+    FieldGetter  getValue;
+    FieldOptions getOptions;
+    FieldSetter  setValue;
+  };
```
Comment (tobi01001) [inline esp8266-fastled-webserver/Fields.cpp:330]: I do believe that for memory and performance reasons, one could change all these strings (except name and label) to uint32_t (FieldGetter / Fieldsetter, maybe min and max...).

The only place where the fieldvalues are required as String is a webserver response. This could either work with the actual value directly or converison to String in the callback function.

The label uses a "string getter" (first entry in Fields) but there is certainly a workaround possible for that one as it is the only label.
The "legacy" solidColor beign returned as comma-separated rgb, the 32bit (24 bit) representation can be used which would require a small adoption the the web app as well.

File: esp8266-fastled-webserver/Fields.cpp [hunk 3]
```diff
@@ -17,479 +17,526 @@
...
+  // convert to seconds
+  utcOffsetInSeconds = tmp * 60;
+
+  Serial.print(F("utcOffsetIndex: ")); Serial.println(utcOffsetIndex);
+  Serial.print(F("utcOffsetInSeconds: ")); Serial.println(utcOffsetInSeconds);
+  timeClient.setTimeOffset(utcOffsetInSeconds);
+  writeAndCommitSettings();
+  return utcOffsetIndex;
+}
+
```
Comment (tobi01001) [inline esp8266-fastled-webserver/Fields.cpp:522]: I do believe this should not go into the Fields.hpp file but somewhere else. The `writeAndCommitSetting()` is currently only used in the main "sketch"...

File: esp8266-fastled-webserver/Fields.cpp [hunk 4]
```diff
@@ -17,479 +17,526 @@
...
+    return (
+      (t == Field_t::String   ) ||
+      (t == Field_t::Label    ) ||
+      (t == Field_t::Color    ) ||
+      (t == Field_t::Boolean  ) ||
+      (t == Field_t::Number   ) ||
+      (t == Field_t::Section  ) ||
+      (t == Field_t::Select   ) ||
+      (t == Field_t::UtcOffset) );
+  }
```
Comment (tobi01001) [inline esp8266-fastled-webserver/Fields.cpp:43]: if you would add a `Invalid` Field at the end, you could change the IsValid function to

```suggestion
constexpr bool IsValid(const Field_t& t) {
    return (t < Field_t::Invalid);
}
```

File: esp8266-fastled-webserver/Fields.cpp [hunk 5]
```diff
@@ -17,479 +17,526 @@
...
+  enum struct Field_t : uint8_t {
+    Number,
+    Boolean,
+    Select,
+    Color,
+    Section,
+    String,
+    Label,
+    UtcOffset,
+  };
```
Comment (tobi01001) [inline esp8266-fastled-webserver/Fields.cpp:32]:
```suggestion
    Label,
    UtcOffset,
    
    Invalid,
  };
```

File: esp8266-fastled-webserver/Fields.cpp [hunk 6]
```diff
@@ -17,479 +17,526 @@
...
+      (t == Field_t::String)    ?  F("String")   :
+      (t == Field_t::Label)     ?  F("Label")    :
+      (t == Field_t::Color)     ?  F("Color")    :
+      (t == Field_t::Boolean)   ?  F("Boolean")  :
+      (t == Field_t::Number)    ?  F("Number")   :
+      (t == Field_t::Section)   ?  F("Section")  :
+      (t == Field_t::Select)    ?  F("Select")   :
+      (t == Field_t::UtcOffset) ?  F("UtcOffsetIndex") :
+                                   F("Invalid")  ;
   }
```
Comment (tobi01001) [inline esp8266-fastled-webserver/Fields.cpp:56]: One could easily live without all these strings by adding (quite) the same enumeration to the **[app.js:](https://github.com/jasoncoon/esp8266-fastled-webserver/blob/b1dd8e9938eca5a93f8d11b1a8bf0389d653b3e6/esp8266-fastled-webserver/data/js/app.js)**
```

var postColorTimer = {};
var postValueTimer = {};

var ignoreColorChange = false;

const fieldtype = Object.freeze({
  Number     : 0,
  Boolean    : 1,
  Select     : 2,
  Color      : 3,
  Section    : 4,
  String     : 5,
  Label      : 6,
  UtcOffset  : 7,
  Invalid    : 8
});

// var ws = new ReconnectingWebSocket("ws://" + address + ":81/", ["arduino"]);

...

$.get(urlBase + "all", function(data) {
      $("#status").html("Loading, please wait...");

      allData = data;

      $.each(data, function(index, field) {
        if (field.type == fieldtype.Number) {
          addNumberField(field);
        } else if (field.type == fieldtype.Boolean) {
          addBooleanField(field);
        } else if (field.type == fieldtype.Select) {
          addSelectField(field);
        } else if (field.type == fieldtype.Color) {
          addColorFieldPalette(field);
          addColorFieldPicker(field);
        } else if (field.type == fieldtype.Select) {
          addSectionField(field);
        } else if (field.type == fieldtype.String) {
          addStringField(field, false);
        } else if (field.type == fieldtype.Label) {
          addStringField(field, true);
        } else if (field.type == fieldtype.UtcOffset) {
          addUtcOffsetIndexField(field, true);
        }
      });

      $(".minicolors").minicolors({
        theme: "bootstrap",
        changeDelay: 200,
        control: "wheel",
        format: "rgb",
        inline: true
      });

...
```

File: esp8266-fastled-webserver/Fields.cpp [hunk 7]
```diff
@@ -17,479 +17,526 @@
...
       }
     }
-
-    if (field.type == NumberFieldType || field.type == UtcOffsetIndexFieldType) {
-      json += ",\"min\":" + String(field.min);
-      json += ",\"max\":" + String(field.max);
+    if ((field.type == Field_t::Number) || (field.type == Field_t::UtcOffset)) {
+      dst[F("min")] = field.min;
+      dst[F("max")] = field.max;
+    }
```
Comment (tobi01001) [inline esp8266-fastled-webserver/Fields.cpp:367]: could be handled in the switch directly?
```
        case Field_t::Boolean:
        case Field_t::Section:
        case Field_t::Select:
            dst[F("value")] = field.getValue(); 
            break;
        case Field_t::UtcOffset:
        case Field_t::Number:
          dst[F("value")] = field.getValue(); 
          dst[F("min")] = field.min;
          dst[F("max")] = field.max;
          break;

```

**Review by tobi01001 — COMMENTED** [review]
Hi,

just a few comments on the proposed changes. Nothing critical though, more like improvmentes / suggestions.

BR

File: esp8266-fastled-webserver/Info.cpp [hunk 8]
```diff
@@ -12,38 +12,110 @@ String WiFi_SSID(bool persistent) {
...
 
-  return json;
+  // convert to String
+  String result;
+  // avoid heap fragmentation by pre-allocating 1023 bytes for result string.
+  // As of 2021-11-21, maximum 890 chars needed for minified result.
+  // As of 2021-11-21, maximum 1031 characters needed for prettified result.
+  result.reserve(1023); 
+  serializeJson(jsonDoc, result);
+  return result;
```
[the remark this answers is not in the document]
Comment (henrygab) [reply esp8266-fastled-webserver/Info.cpp:120]: Hi Tobi,

Yes, there is more that can be done!  More passing of `jsonDoc` references is likely to come.  These are smaller steps towards this.

for example, I was thinking of moving `jsonDoc` up, so that failures in `jsonDoc` creation / filling could return an error code.

This is an active area for me right now, so expect more in the near future.

File: esp8266-fastled-webserver/Fields.cpp (hunk 1 above)
Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp:343]: Yes, if this was to be done.  Currently, I am trying to keep compatibility, in case folks have relied upon this JSON output (e.g., using their own parser) to display the state of the webserver.  If I were to change the type, this would break anything existing code expecting the stringified version.  Thus, I decided against any breaking changes at this time.

Hope that makes sense?

Comment (tobi01001) [reply esp8266-fastled-webserver/Fields.cpp:343]: For sure certainly does make sense.
I just doubt that there are actually many (or any) people relying on the "/all" output apart from the web app itself.
I do but I also completely changed the project for my personal taste and use....

And by introducing a utc offset  as additional field with explicit values being required on the web app, compatibility is somewha broken anyway ;-)

File: esp8266-fastled-webserver/Fields.cpp (hunk 5 above)
Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp:32]: While some designs do include an explicit "invalid" value for an enumeration, there are definitely reasons to avoid this (at least with scoped enumerations).

One benefit of scoped enumerations (`enum struct ...` or `enum class`) is that, when writing a `switch` statement, the compiler will generate a warning when there is a value that is not handled by a `case` statement.

In fact, when adding the UtcOffset, I avoided a bug by this exact warning.

If adding an explicit 'invalid' value, then it might suggest that value is the only invalid value (which is not the case).  Moreover, for this type, there should *never* be an invalid value.

Again, I realize this is a common pattern.  I did consider it, but given the enum's purposes and use model, I explicitly chose not to add it here.

File: esp8266-fastled-webserver/Fields.cpp (hunk 4 above)
Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp]: First, **_THANK YOU_** for the in-depth review!  It's clear that you have spent some time as a developer, and I really appreciate each of these. 
Even if I decide differently, I greatly appreciate the questions so that the alternatives are considered.  Awesome!

I think I understand the design pattern you are suggesting.

One benefit of your design pattern is that it automatically updates when a new enum type is added.  Some drawbacks of your design pattern are that `Invalid` enum values are allowed to exist, it has an unwritten assumption that the enum values are all sequential, a case for `Invalid` must be added to all switch statements, and there's a subtle implication that there is only one invalid value.

FYI, the compiler will optimize the above to a simple check of `< 8`, so the underlying code remains identical.  If we were targeting C++14 or later, this function would be written as a case statement (to automatically warn for missing values).

Both design patterns have benefits and drawbacks, so without a stronger reason, I would prefer to stick with the current function.

Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp]: Oh, but *do* re-open if you feel I should reconsider, and can help me understand additional aspects to consider!

File: esp8266-fastled-webserver/Fields.cpp (hunk 6 above)
Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp:56]: Yes, you are right.

This would change the JSON data generated.
I want to avoid that, in case anyone wrote scripts that interpret this JSON.  Changing the underlying data type to an `enum` would break those scripts.  Thus, for at least this PR, I am choosing to avoid breaking changes in the JSON data.

File: esp8266-fastled-webserver/Fields.cpp (hunk 2 above)
Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp:330]: You are **_absolutely_** correct.  See prior responses ... explicit choice not to change JSON format with this PR.

File: esp8266-fastled-webserver/Fields.cpp (hunk 7 above)
Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp:367]: You are correct.  

There's some oddity with the use of flash-based strings, where they do not get optimized to single storage in flash.  (e.g., there will be two copies of "value" in flash just from the quoted code snippet).  However, I agree that transitioning to a per-type serialization, even if slightly larger resulting binary, would be worth the clarity it adds.

Can you please open a new issue to have this changed, so that each field type handles its own serialization?

File: esp8266-fastled-webserver/Fields.cpp (hunk 3 above)
Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp:522]: Yes, I agree.  In fact, I would argue that all the get/set functions should be moved outside this file, as well as the variables themselves.

Would you open an issue to move the getter/setter functions into their own file?

**Comment by henrygab** [comment]
@tobi01001 -- Awesome review, I really appreciate your ability to quickly pinpoint those additional areas for improvement!  I hope you'll open a couple issues (so that you remain the owner of the issue) to track the noted enhancements, as they're great ideas.  Thank you!

File: esp8266-fastled-webserver/Fields.cpp (hunk 7 above)
Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp:367]: opened #219 to track this.

File: esp8266-fastled-webserver/Fields.cpp (hunk 3 above)
Comment (henrygab) [reply esp8266-fastled-webserver/Fields.cpp:522]: Now tracked in #220.

**Comment by henrygab** [comment]
Hi @jasoncoon -- Are you good with these changes?

**Review by jasoncoon — APPROVED** [review]

**Comment by jasoncoon** [comment]
I'm working on adding Fibonacci1024, and the `/all` response is being truncated. I assume I'll need to adjust the size allocated to the jsonDoc?  Would it make sense to split this into multiple API calls? Perhaps one call to enumerate the fields, including only the basic info, then one additional request per select field to fully describe the options?  Since the bulk of the data is the list of pattern and palette names.

**Comment by henrygab** [comment]
Hi Jason,

I'm interested!  Please message me on Tindie with details on expected costs, etc?

You can use [ArduinoJSON Assistant](https://arduinojson.org/v6/assistant/) to help figure out the size of the buffer needed.  I entered the longest possible strings in each field, then added ~100 bytes.  Looks like I didn't add enough.

Would it make sense to split the API?  Yes, at least for pattern / palette lists.

Maybe define a new data type, which is a string representing a relative URI, where the caller calls that URI as a REST API to get the additional data?

I'm sure there are "correct" ways to do this in REST, but I'm not deeply familiar with how REST APIs normally self-document....

**Comment by tobi01001** [comment]
I do believe that this is good but maybe not neccessary ;-)

The /all API is only used to fill the web app with the configuration / parameters.
The only dynamic part is in the current values.

So rather than increasing the JSON Buffer I propose to:
- decouple fixed values (i.e. the configuration) from the dynamic / changing values
- /all only serves the configuration
- /allvalues serves current values of parameters

/all is fixed as long as there is no change in number in the software affecting the fields and its contents. So why not letting the ESP app writing the config to a file on the LittleFS (e.g. config_all.json).
On SW update/magic number change (I would tend to use a CRC which is set to 0 on SW updates), the application could in `setup()` rewrite the file to the current config.

And all the app would have to do with the API call to `/all` is to serve that particular file instead of the StaticJSONBuffer...

**Comment by henrygab** [comment]
Caveats which Jason should confirm, but I will presume as true for this post:

<details><summary> It's OK to break compatibilty withe old REST APIs.</summary><P/>

If yes, this means it's OK to break compatibility if a 3rd party had used the REST APIs, such as using custom `curl` scripts with `crond`, etc.  In essence, the only supported client is the on-device Web UI, so that updates needn't consider backwards-compatibility.  I think Jason will be OK with this?

</details>
<details><summary> It's OK to take a permanent dependent on having LittleFS storage.</summary><P/>

If yes, this means boards that don't have LittleFS (or SPIFFS) support will not work.  I don't foresee this being a concern, but it's better to be sure.

</details><HR/>

### @tobi01001's solution is excellent

In fact, if Jason confirms those are OK, then it'll be time to drop the "EEPROM" code altogether, in favor of reading/writing the parameters through LittleFS, using ArduinoJson.

**Comment by jasoncoon** [comment]
It's definitely OK with me to break compatibility with old REST APIs.  Might need to check in with @balloob to see if there's a good way to prevent breaking the HomeAssistant integration: https://github.com/home-assistant/core/tree/dev/homeassistant/components/evil_genius_labs

Permanent LittleFS dependency is also fine with me.

It'd be great if this could be done in a way that wouldn't require having to update multiple places in the code and/or file when new patterns/palettes are added, updated, renamed, etc.

I also think the patterns and palettes list should probably be separate requests, as they are each so large.

**Comment by balloob** [comment]
Can we add a version number to the API so I can differentiate the different APIs?

**Comment by tobi01001** [comment]
> 
@henrygab 
I actually did not mean the EEPROM settings... Personally I would avoid writing them to a json file in the file system. 
As they are being written quite frequently it may / will cause quite a lot of additional wear to the flash - unless you write and read only binary data. but That is then the same just with the overhead for the file system ;-)

For easier maintenance and for the sake of reliability I would suggest that (modifiable) settings to be stored are managed in a structure like e.g.:

<details><summary>structure</summary>

```cpp
typedef struct {
  uint8_t settingsMagicByte = SETTINGS_MAGIC_BYTE;
  CRGB solidColor = CRGB::Blue;
  uint8_t speed = 30;
  uint8_t power = 1;
  uint8_t brightness = brightnessMap[brightnessIndex];
  uint8_t cooling = 49;
  uint8_t sparking = 60;
  uint8_t currentPatternIndex = DEFAULT_PATTERN_INDEX; // Index number of which pattern is current
  uint8_t autoplay = 0;
  uint8_t autoplayDuration = 10;
  uint8_t showClock = 0;
  uint8_t clockBackgroundFade = 160;
  uint8_t utcOffsetIndex = 24; 
  uint8_t currentPaletteIndex = 0;
  uint8_t twinkleSpeed = 4;
  uint8_t twinkleDensity = 5;
  uint8_t coolLikeIncandescent = 1;
} strEEPROMSettings;
```
</details>

This way, the settings management incl. EEPROM write and read is just:

<details><summary>Handling</summary>

```cpp
strEEPROMSettings EEPROMSettings;
...
EEPROM.begin(sizeof(EEPROMSettings))
...
void readSettings()
{
  // check for "magic number" so we know settings have been written to EEPROM
  // and it's not just full of random bytes
  if (EEPROM.read(0) != SETTINGS_MAGIC_BYTE) {
    return;
  }
  EEPROM.get(0, EEPROMSettings);
  setUtcOffsetIndex(EEPROMSettings.utcOffsetIndex);
}

void writeAndCommitSettings() {
  EEPROMSettings.settingsMagicByte = SETTINGS_MAGIC_BYTE;
  EEPROM.put(0, EEPROMSettings);
  EEPROM.commit();
}
```

</details>

The EEPROM uses a reserved section in the flash where a complete sector needs to be erased and rewritten. Reliability could further be increased by using e.g. [EEPROM_Rotate](https://github.com/xoseperez/eeprom_rotate) which is basically a drop-in replacement of the EEPROM class and which will cycle through different sectors...

Long story short: As the EEPROM is only emulated in the flash, I see no benefit (other than maybe readability and code maintenance) to switch to LittleFS with file sytem overhead for the settings to be saved and reread quite often...

@jasoncoon If Fields (w/o values) are all written / stored in a file, I think there is no need to separate the "options" for patterns and palettes - unless one would only need to read those. But this could go into a separate API call. The webserver does serve all the other files (which are bigger in size) wihtout issues and so it can also serve this one easily (and quick).

**Comment by jasoncoon** [comment]
@tobi01001 just so I'm clear, we're talking about the fields (w/o values) being written/stored in a file automatically/dynamically, in setup, on startup? Not maintained separately in a static file that would need to be kept in sync with the patterns, palettes, etc in code?

**Comment by tobi01001** [comment]
> 

Yes, exactly. 
A file which is generated from the fields itself in case the configuration changed (or the file is not yet available).

I did this already but as I am personally on ArduinoJson V5 and AsyncWebserver (and very limited time currently) I cannot come up with a PR on the fly... ;-) 

In my case with AsyncWebServer this would be as easy as:

<details><summary>Update File</summary>

```cpp
void getAllJSON(JsonArray &arr)
{
  for (uint8_t i = 0; i < getFieldCount(); i++)
  {
    Field field = fields[i];
    JsonObject& obj = arr.createNestedObject();
    obj[F("name")]  = field.name;
    obj[F("label")] = field.label;
    obj[F("type")]  = (int)field.type;
    
    if (field.type == NumberFieldType) 
    {
      obj[F("min")] = field.min;
      obj[F("max")] = field.max;
    }

    if (field.getOptions)
    {
      JsonArray &ar = obj.createNestedArray(F("options"));
      field.getOptions(ar);
    }
  }
}

void updateConfigFile(void)
{
  DynamicJsonBuffer buffer;
  JsonArray& root = buffer.createArray();
  
  getAllJSON(root);

  File config_Json = LittleFS.open(F("/config/config_all.json"), "w");

  root.printTo(config_Json);

  config_Json.close();
}

```


</details>

<details><summary>Serve the file</summary>

```cpp

... 

  server.on("/all", HTTP_GET, [](AsyncWebServerRequest *request) {
    if(!LittleFS.exists("/config/config_all.json"))
    {
      updateConfigFile();
    }
    request->send(LittleFS, "/config/config_all.json", "application/json");
  });


  server.on("/allvalues", HTTP_GET, [](AsyncWebServerRequest *request) {
    AsyncJsonResponse * response = new AsyncJsonResponse();

    JsonObject &root = response->getRoot();
    JsonArray &arr = root.createNestedArray("values");
    if(!getAllValuesJSONArray(arr))
    {
      JsonObject &obj = arr.createNestedObject();
      obj[F("ValueError")] = F("Did not read any values!");
    }
    response->setLength();
    request->send(response);
  });
```
</details>

And the app.js needs to perform a second request to e.g /allvalues to update the web app to the current values with something like:

<details><summary>app.js</summary>

```js
  $.get(urlBase + "all", function(data) {

    })
    .fail(function(errorThrown) {
      console.log(errorThrown);
    })
    .done(function(name, value, test) {
	$("#status").html("Done...");
	$.get(urlBase + "allvalues", function(data) {
        $("#status").html("Loading, current values...");
	for(i=0; i<data.values.length; i++) {
	  updateFieldValue( data.values[i].name,  data.values[i].value);
       }
       })
       .fail(function(errorThrown) {
       console.log("error: " + errorThrown);
       })
       .done(function(name, value, test) {
      $("#status").html("Ready");
      });
   });

```
</details>

**Comment by henrygab** [comment]
Having two methods to read/parse/store configuration (EEPROM and JSON) is going to be error-prone, and increases maintenance.  I still think it makes sense to store any user-configurable data in `/config.json`, and to stop using the EEPROM.  Reading / updating the configuration is not a performance-sensitive operation, so I think the benefits in maintainability will outweigh the downsides.

Any objection if I work towards Tobi's solution, ditch all the EEPROM code, and store only in the file system?

**Comment by henrygab** [comment]
OK, I will open a new issue to track that work.  Thanks!

**Comment by henrygab** [comment]
> Can we add a version number to the API so I can differentiate the different APIs?

Yes, that's a fine idea.
